### DISCUSSION TITLE
Collection sort => case insensitive

### DISCUSSION THREAD

**Discussion by peterhijma** [pocketbase/pocketbase#2971] [Q&A] [answered]
Hi! How could I make the sorting case insensitive? Currently if I order on a field, `sort=field`, with these column values:

```
foo
hi
Foo
bar
```
It'll lead to:

```
Foo
bar
foo
hi
```

I'd prefer something like this:

```
bar
foo
Foo
hi
```

Thanks in advance!

**Comment by ganigeorgiev** [comment, marked as answer]
At the moment the web APIs don't support specifying the sorting collation (_I've added in the roadmap to consider it_).

For now, if you want efficiently to sort case insensitive, you could create a helper view with `COLLATE NOCASE` order clause. Something like:

```sql
SELECT id, field1, field2, ...
FROM yourCollection
ORDER BY field1 COLLATE NOCASE ASC
```

**Reply by peterhijma** [reply]
Ah nice! Didn't know about such views yet. Thanks :)

**Comment by peterhijma** [comment]
It works without an api rule (accessible by everyone). But when adding this rule to the view collection, the sorting is wrong again 🤔:

```
@collection.foo.user ?= @request.auth.id
```

What mistake do I make here?

**Reply by ganigeorgiev** [reply]
There are not enough details about your collections structure to understand what's going on.

You can inspect the generated query by starting the executable with `--debug`.

If that doesn't help, could you export your collections configurations and paste it here as JSON (_Admin UI > Settings > Export collections_)?

**Reply by ganigeorgiev** [reply]
I guess it is possible the above `@collection.*` join rule to interfere with the order that is defined in the view because the executed query technically would be something like:

```sql
SELECT yourView.*
FROM yourView
LEFT JOIN foo
WHERE foo.user = {@request.auth.id}
```

**Reply by peterhijma** [reply]
The resulting queries using `--debug`:

```sql
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = '_pb_users_auth_' OR LOWER([[name]])='_pb_users_auth_' LIMIT 1
[0.00ms] SELECT `users`.* FROM `users` WHERE `users`.`id`='sej1ba3aqmuv5cz' LIMIT 1
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = 'foo_sorted' OR LOWER([[name]])='foo_sorted' LIMIT 1

[0.00ms] SELECT COUNT(DISTINCT [[foo_sorted.id]]) FROM `foo_sorted` LEFT JOIN `foo_sorted` `__collection_foo_sorted` WHERE [[__collection_foo_sorted.user]] = 'sej1ba3aqmuv5cz'
[0.00ms] SELECT DISTINCT `foo_sorted`.* FROM `foo_sorted` LEFT JOIN `foo_sorted` `__collection_foo_sorted` WHERE [[__collection_foo_sorted.user]] = 'sej1ba3aqmuv5cz' LIMIT 30
```

So yes, I guess your assumption is right :D

I can't access the request user in the SQL part, I assume?

**Reply by ganigeorgiev** [reply]
> I can't access the request user in the SQL part, I assume?

No, views can't be parameterized.

As a workaround, if `foo` in the above is the same collection that is used in the view, to avoid the rule join clause, you can define a `user` field in the view and check it directly in the rules like `user = @request.auth.id`.

**Comment by peterhijma** [comment]
> > I can't access the request user in the SQL part, I assume?
> 
> No, views can't be parameterized.
> 
> As a workaround, if `foo` in the above is the same collection that is used in the view, to avoid the rule join clause, you can define a `user` field in the view and check it directly in the rules like `user = @request.auth.id`.

Works perfectly in my case! Thanks for the (quick) answers once again 🥳

**Reply by ganigeorgiev** [reply]
Another workaround that I haven't tested but could be more universal is to try to define `(ROW_NUMBER() OVER()) as id` (_or as any other column_) in the view and then for the query you can use the `?sort=id` or `?sort=-id` as part of your list request to order ASC/DESC based on the default view order.

_But in all cases avoiding the extra join should be more performant anyway because if you decide to use `@collection.*` the actual rule will need also another point of reference to match the joined collection rows with the ones from the view, aka. something like `@collection.foo.id ?= id && @collection.foo.user ?= @request.auth.id`._
